<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Information Divergences of Methylation Levels — estimateDivergence • MethylIT</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.1/css/all.min.css" integrity="sha256-PbSmjxuVAzJ6FPvNYsrXygfGhNJYyZ2GktDbkMBqQZg=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.1/css/v4-shims.min.css" integrity="sha256-A6jcAdwFD48VMjlI3GDxUd+eCQa7/KWy6G9oe/ovaPA=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Information Divergences of Methylation Levels — estimateDivergence" />
<meta property="og:description" content="This function prepares the data for the estimation of
information divergences and works as a wrapper calling the functions that
compute selected information divergences of methylation levels. In the
downstream analysis, the probability distribution of a given information
divergence is used in Methyl-IT as the null hypothesis of the noise
distribution, which permits, in a further signal detection step, the
discrimination of the methylation regulatory signal from the background
noise.For the current version, two information divergences of methylation
levels are computed by default: 1) Hellinger divergence (\emph{H}) and 2)
the total variation distance (\emph{TVD}). In the context of methylation
analysis \emph{TVD} corresponds to the absolute difference of methylation
levels. Here, although the variable reported is the total variation
(\emph{TV}), the variable actually used for the downstream analysis is
\emph{TVD}. Once a differentially methylated position (DMP) is identified
in the downstream analysis, \emph{TV} is the standard indicator of
whether the cytosine position is hyper- or hypo-methylated.

The option to compute the J-information divergence (JD) is currently
provided. The motivation to introduce this divergence is given in the
help of function \code{\link{estimateJDiv}}.

" />
<meta property="og:image" content="/logo.png" />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MethylIT</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/MethylIT.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/cancer_example.html">Example of Methylome Analysis with MethylIT using Cancer Datasets</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Information Divergences of Methylation Levels</h1>
    
    <div class="hidden name"><code>estimateDivergence.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function prepares the data for the estimation of
information divergences and works as a wrapper calling the functions that
compute selected information divergences of methylation levels. In the
downstream analysis, the probability distribution of a given information
divergence is used in Methyl-IT as the null hypothesis of the noise
distribution, which permits, in a further signal detection step, the
discrimination of the methylation regulatory signal from the background
noise.</p><pre>For the current version, two information divergences of methylation
levels are computed by default: 1) Hellinger divergence (\emph{H}) and 2)
the total variation distance (\emph{TVD}). In the context of methylation
analysis \emph{TVD} corresponds to the absolute difference of methylation
levels. Here, although the variable reported is the total variation
(\emph{TV}), the variable actually used for the downstream analysis is
\emph{TVD}. Once a differentially methylated position (DMP) is identified
in the downstream analysis, \emph{TV} is the standard indicator of
whether the cytosine position is hyper- or hypo-methylated.

The option to compute the J-information divergence (JD) is currently
provided. The motivation to introduce this divergence is given in the
help of function \code{\link{estimateJDiv}}.
</pre>

    </div>

    <pre class="usage"><span class='fu'>estimateDivergence</span>(
  <span class='no'>ref</span>,
  <span class='no'>indiv</span>,
  <span class='kw'>Bayesian</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>columns</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>min.coverage</span> <span class='kw'>=</span> <span class='fl'>4</span>,
  <span class='kw'>min.meth</span> <span class='kw'>=</span> <span class='fl'>4</span>,
  <span class='kw'>min.umeth</span> <span class='kw'>=</span> <span class='fl'>0</span>,
  <span class='kw'>high.coverage</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>percentile</span> <span class='kw'>=</span> <span class='fl'>0.999</span>,
  <span class='kw'>JD</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>num.cores</span> <span class='kw'>=</span> <span class='fl'>1L</span>,
  <span class='kw'>tasks</span> <span class='kw'>=</span> <span class='fl'>0L</span>,
  <span class='kw'>meth.level</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>logbase</span> <span class='kw'>=</span> <span class='fl'>2</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='no'>...</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>ref</th>
      <td><p>The GRanges object of the reference individual that will be used
in the estimation of the information divergence.</p></td>
    </tr>
    <tr>
      <th>indiv</th>
      <td><p>A list of GRanges objects from the individuals that will be
used in the estimation of the information divergence.</p></td>
    </tr>
    <tr>
      <th>Bayesian</th>
      <td><p>Logical. Whether to perform the estimations based on
posterior estimations of methylation levels.</p></td>
    </tr>
    <tr>
      <th>columns</th>
      <td><p>Vector of one or two integer numbers denoting the indexes of
the columns where the methylated and unmethylated read counts are found
or, if meth.level = TRUE, the columns corresponding to the methylation
levels. If columns = NULL and meth.level = FALSE, then columns = c(1,2)
is assumed. If columns = NULL and meth.level = TRUE, then columns = 1 is
assumed.</p></td>
    </tr>
    <tr>
      <th>min.coverage</th>
      <td><p>An integer or an integer vector of length 2. Cytosine
sites where the coverage in both samples, 'x' and 'y', are less than
'min.coverage' are discarded. The cytosine site is preserved, however, if
the coverage is greater than 'min.coverage' in at least one sample. If
'min.coverage' is an integer vector, then the corresponding min coverage
is applied to each sample.</p></td>
    </tr>
    <tr>
      <th>min.meth</th>
      <td><p>An integer or an integer vector of length 2. Cytosine sites
where the numbers of read counts of methylated cytosine in both samples,
'1' and '2', are less than 'min.meth' are discarded. If 'min.meth' is an
integer vector, then the corresponding min number of reads is applied to
each sample. Default is min.meth = 4.</p></td>
    </tr>
    <tr>
      <th>min.umeth</th>
      <td><p>An integer or an integer vector of length 2
(\(min.umeth = c(min.umeth1, min.umeth2)\)). Min number of
reads to consider cytosine position. Specifically cytosine positions
where (uC &lt;= min.umeth) &amp; (mC &gt; 0) &amp; (mC &lt;= min.meth) hold will be
removed, where mC and uC stand for the numbers of methylated and
unmethylated reads. Default is min.umeth = 0.</p></td>
    </tr>
    <tr>
      <th>high.coverage</th>
      <td><p>An integer for read counts. Cytosine sites having
higher coverage than this are discarded.</p></td>
    </tr>
    <tr>
      <th>percentile</th>
      <td><p>Threshold to remove the outliers from each file and all
files stacked.</p></td>
    </tr>
    <tr>
      <th>JD</th>
      <td><p>Logic (Default:FALSE). Option on whether to add a column with
values of J-information divergence (see <code><a href='estimateJDiv.html'>estimateJDiv</a></code>).
It is only compute if JD = TRUE and meth.level = FALSE.</p></td>
    </tr>
    <tr>
      <th>num.cores</th>
      <td><p>The number of cores to use, i.e. at most how many child
processes will be run simultaneously (see 'bplapply' function from
BiocParallel package).</p></td>
    </tr>
    <tr>
      <th>tasks</th>
      <td><p>integer(1). The number of tasks per job. value must be a scalar
integer &gt;= 0L. In this documentation a job is defined as a single call
to a function, such as bplapply, bpmapply etc. A task is the division of
the X argument into chunks. When tasks == 0 (default), X is divided as
evenly as possible over the number of workers (see MulticoreParam from
BiocParallel package).</p></td>
    </tr>
    <tr>
      <th>meth.level</th>
      <td><p>Logic. Whether methylation levels are given in place of
counts. Default is FALSE.</p></td>
    </tr>
    <tr>
      <th>logbase</th>
      <td><p>Logarithm base used to compute the JD (if JD = TRUE).
Logarithm base 2 is used as default (bit unit). Use
logbase = \(exp(1)\) for natural logarithm.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>if TRUE, prints the function log to stdout</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional parameters for 'uniqueGRanges' function.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>An object from "infDiv" class with the four columns of counts, the
information divergence, and additional columns: 1) The original matrix
of methylated (c_i) and unmethylated (t_i) read counts from control
(i=1) and treatment (i=2) samples. 2) p1" and "p2": methylation levels
for control and treatment, respectively. 3) "bay.TV": total variation
TV = p2 - p1. 4) "TV": total variation based on simple counts:
TV=c1/(c1+t1)-c2/(c2+t2). 5) "hdiv": Hellinger divergence. If
Bayesian = TRUE, the results are based on the posterior estimations of
methylation levels.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>If read counts are provided, then Hellinger divergence is computed
as given in the first formula from Theorem 1 from reference 1. In the
present case:</p><pre>\deqn{H = 2*(n[1] + 1)*(n[2] + 1)*((sqrt(p[1]) - sqrt(p[2]))^2 +
         (sqrt(1-p[1]) - sqrt(1-p[2]))^2)/(n[1] + n[2] + 2)}

where \eqn{n[1]} and \eqn{n[2]} are the coverage for the control and
treatment, respectively. Notice that each row from the matrix of counts
correspond to a single cytosine position and has four values
corresponding to "mC1" and "uC1" (control), and mC2" and "uC2" for
treatment.

According with the above equation, to estimate Hellinger divergence, not
only the methylation levels are considered in the estimation of H,
but also the control and treatment coverage at each given cytosine site.
At this point, it is worthy to do mention that if the reference sample is
derived with function \code{\link{poolFromGRlist}} using the 'sum' of
read counts to conpute a methylation pool, then 'min.coverage' parameter
value must be used to prevent an over estimation of the divergence for
low coverage cytosines sites. For example, if a reference sample is
derived as the methylation pool of read count sum from 3 individuals and
we want to consider only methylation sites with minimum coverage of 4,
then we can set min.coverage = c(12, 4), where the number 12 (3 x 4) is
the minimum coverage requested for the each cytosine site in the
reference sample.

If the methylation levels are provided in place of counts, then the
Hellinger divergence is computed as:
\deqn{H = (sqrt(p[1]) - sqrt(p[2]))^2 + (sqrt(1 - p[1]) -
          sqrt(1 - p[2]))^2}

This formula assumes that the probability vectors derived from the
methylation levels (p_ij) p_j = c(p_ij, 1 - p_ij) (see function
'estimateHellingerDiv') are an unbiased estimation of the expected one.
The function applies a pairwise filtering after building a single GRanges
from the two GRanges objects. Experimentally available cytosine sites are
paired using the function 'uniqueGRanges'.

It is important to observe that several filtering conditions are provided
to select biological meaningful cytosine positions, which prevent to
carry experimental errors in the dowstream analyses. By filtering the
read count we try to remove bad quality data, which would be in the edge
of the experimental error originated by the BS-seq sequencing. It is
responsability of the user to check whether cytosine positions used in
the analysis are biological meaningful. For example, a cytosine position
with counts mC1 = 10 and uC1 = 20 in the 'ref' sample and mC2 = 1 &amp; uC2 =
0 in an 'indv' sample will lead to methylation levels p1 = 0.333 and p2 =
1, respectively, and TV = p2 - p1 = 0.667, which apparently indicates a
hypermethylated site. However, there are not enough reads supporting p2 =
1. A Bayesian estimation of TV will reveal that this site would be, in
fact, hypomethylated. So, the best practice will be the removing of sites
like that. This particular case is removed under the default settings:
min.coverage = 4, min.meth = 4, and min.umeth = 0 (see example for
function \code{\link{uniqueGRfilterByCov}}, called by
estimateDivergence).
</pre>


    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'>## The read count data are created</span>
    <span class='no'>num.samples</span> <span class='kw'>&lt;-</span> <span class='fl'>250</span>
    <span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span>(<span class='kw'>chr</span> <span class='kw'>=</span> <span class='st'>"chr1"</span>, <span class='kw'>start</span> <span class='kw'>=</span> <span class='fl'>1</span>:<span class='no'>num.samples</span>,
                    <span class='kw'>end</span> <span class='kw'>=</span> <span class='fl'>1</span>:<span class='no'>num.samples</span>,<span class='kw'>strand</span> <span class='kw'>=</span> <span class='st'>'*'</span>,
                    <span class='kw'>mC</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/stats/NegBinomial.html'>rnbinom</a></span>(<span class='kw'>size</span> <span class='kw'>=</span> <span class='no'>num.samples</span>, <span class='kw'>mu</span> <span class='kw'>=</span> <span class='fl'>4</span>, <span class='kw'>n</span> <span class='kw'>=</span> <span class='fl'>500</span>),
                    <span class='kw'>uC</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/stats/NegBinomial.html'>rnbinom</a></span>(<span class='kw'>size</span> <span class='kw'>=</span> <span class='no'>num.samples</span>, <span class='kw'>mu</span> <span class='kw'>=</span> <span class='fl'>4</span>, <span class='kw'>n</span> <span class='kw'>=</span> <span class='fl'>500</span>))
    <span class='no'>y</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/data.frame.html'>data.frame</a></span>(<span class='kw'>chr</span> <span class='kw'>=</span> <span class='st'>"chr1"</span>, <span class='kw'>start</span> <span class='kw'>=</span> <span class='fl'>1</span>:<span class='no'>num.samples</span>,
                    <span class='kw'>end</span> <span class='kw'>=</span> <span class='fl'>1</span>:<span class='no'>num.samples</span>, <span class='kw'>strand</span> <span class='kw'>=</span> <span class='st'>'*'</span>,
                    <span class='kw'>mC</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/stats/NegBinomial.html'>rnbinom</a></span>(<span class='kw'>size</span> <span class='kw'>=</span> <span class='no'>num.samples</span>, <span class='kw'>mu</span> <span class='kw'>=</span> <span class='fl'>4</span>, <span class='kw'>n</span> <span class='kw'>=</span> <span class='fl'>500</span>),
                    <span class='kw'>uC</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/stats/NegBinomial.html'>rnbinom</a></span>(<span class='kw'>size</span> <span class='kw'>=</span> <span class='no'>num.samples</span>, <span class='kw'>mu</span> <span class='kw'>=</span> <span class='fl'>4</span>, <span class='kw'>n</span> <span class='kw'>=</span> <span class='fl'>500</span>))
    <span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>makeGRangesFromDataFrame</span>(<span class='no'>x</span>, <span class='kw'>keep.extra.columns</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
    <span class='no'>y</span> <span class='kw'>&lt;-</span> <span class='fu'>makeGRangesFromDataFrame</span>(<span class='no'>y</span>, <span class='kw'>keep.extra.columns</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
    <span class='no'>hd</span> <span class='kw'>&lt;-</span> <span class='fu'>estimateDivergence</span>(<span class='kw'>ref</span> <span class='kw'>=</span> <span class='no'>x</span>, <span class='kw'>indiv</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span>(<span class='no'>y</span>), <span class='kw'>JD</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)<span class='kw'>[[</span><span class='fl'>1</span>]]</div><div class='output co'>#&gt; <span class='message'>*** Processing sample #1 </span></div><div class='output co'>#&gt; <span class='message'> *** Building coordinates for the new GRanges object ...</span></div><div class='output co'>#&gt; <span class='message'> *** Strand information is preserved ...</span></div><div class='output co'>#&gt; <span class='message'> *** Processing GRanges for sample: #1</span></div><div class='output co'>#&gt; <span class='message'> *** Processing GRanges for sample: #2...</span></div><div class='output co'>#&gt; <span class='message'> *** Sorting by chromosomes and start positions...</span></div><div class='output co'>#&gt; *** Estimating Hellinger divergence... 
#&gt; * Coercing from list to vector...</div><div class='input'>
<span class='co'>## Keep in mind that Hellinger and J divergences are, in general, correlated!</span>
    <span class='fu'><a href='https://rdrr.io/r/stats/cor.test.html'>cor.test</a></span>(<span class='kw'>x</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/numeric.html'>as.numeric</a></span>(<span class='no'>hd</span>$<span class='no'>hdiv</span>), <span class='kw'>y</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/numeric.html'>as.numeric</a></span>(<span class='no'>hd</span>$<span class='no'>jdiv</span>),
            <span class='kw'>method</span> <span class='kw'>=</span> <span class='st'>"kendall"</span>)</div><div class='output co'>#&gt; 
#&gt; 	Kendall's rank correlation tau
#&gt; 
#&gt; data:  as.numeric(hd$hdiv) and as.numeric(hd$jdiv)
#&gt; z = 19.516, p-value &lt; 2.2e-16
#&gt; alternative hypothesis: true tau is not equal to 0
#&gt; sample estimates:
#&gt;       tau 
#&gt; 0.9136781 
#&gt; </div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#details">Details</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Robersy Sanchez</p>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Robersy Sanchez.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


