<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Evaluate DMPs Classification — evaluateDIMPclass • MethylIT</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.1/css/all.min.css" integrity="sha256-PbSmjxuVAzJ6FPvNYsrXygfGhNJYyZ2GktDbkMBqQZg=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.1/css/v4-shims.min.css" integrity="sha256-A6jcAdwFD48VMjlI3GDxUd+eCQa7/KWy6G9oe/ovaPA=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Evaluate DMPs Classification — evaluateDIMPclass" />
<meta property="og:description" content="For a given cutpoint (previously estimated with the function
estimateCutPoint), 'evaluateDIMPclass' will return the evaluation of the
classification of DMPs into two clases: DMPS from control and DMPs
from treatment samples." />
<meta property="og:image" content="/logo.png" />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MethylIT</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/MethylIT.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/cancer_example.html">Example of Methylome Analysis with MethylIT using Cancer Datasets</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/">News</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Evaluate DMPs Classification</h1>
    
    <div class="hidden name"><code>evaluateDIMPclass.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>For a given cutpoint (previously estimated with the function
estimateCutPoint), 'evaluateDIMPclass' will return the evaluation of the
classification of DMPs into two clases: DMPS from control and DMPs
from treatment samples.</p>
    </div>

    <pre class="usage"><span class='fu'>evaluateDIMPclass</span>(
  <span class='no'>LR</span>,
  <span class='no'>control.names</span>,
  <span class='no'>treatment.names</span>,
  <span class='kw'>column</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='kw'>hdiv</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>TV</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>bay.TV</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>wprob</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>pos</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>),
  <span class='kw'>classifier</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"logistic"</span>, <span class='st'>"pca.logistic"</span>, <span class='st'>"lda"</span>, <span class='st'>"qda"</span>, <span class='st'>"pca.lda"</span>, <span class='st'>"pca.qda"</span>),
  <span class='kw'>pval.col</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>n.pc</span> <span class='kw'>=</span> <span class='fl'>1</span>,
  <span class='kw'>center</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>scale</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>interaction</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>output</span> <span class='kw'>=</span> <span class='st'>"conf.mat"</span>,
  <span class='kw'>prop</span> <span class='kw'>=</span> <span class='fl'>0.6</span>,
  <span class='kw'>num.boot</span> <span class='kw'>=</span> <span class='fl'>100</span>,
  <span class='kw'>num.cores</span> <span class='kw'>=</span> <span class='fl'>1L</span>,
  <span class='kw'>tasks</span> <span class='kw'>=</span> <span class='fl'>0L</span>,
  <span class='kw'>seed</span> <span class='kw'>=</span> <span class='fl'>1234</span>,
  <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>LR</th>
      <td><p>An object from 'pDMP' class. including control and treatment
GRanges containing divergence values for each DMP in the meta-column. LR
is generated by the function 'selectDIMP' Each GRanges object must
correspond to a sample. For example, if  a sample is named 's1', then
this sample can be accessed in the list of GRanges objects as LR$s1.</p></td>
    </tr>
    <tr>
      <th>control.names</th>
      <td><p>Names/IDs of the control samples, which must be include
in the variable LR.</p></td>
    </tr>
    <tr>
      <th>treatment.names</th>
      <td><p>Names/IDs of the treatment samples, which must be
included in the variable LR.</p></td>
    </tr>
    <tr>
      <th>column</th>
      <td><p>a logical vector for column names for the predictor variables
to be used: Hellinger divergence "hdiv", total variation "TV", TV
estimated with Bayesian correction of methylation levels "bay.TV",
probability of potential DMP "wprob", and the relative cytosine site
position "pos" in respect to the chromosome where it is located. The
relative position is estimated as (x - x.min)/(x.max - x), where x.min
and x.max are the maximum and minimum for the corresponding chromosome,
respectively. If "wprob = TRUE", then Logarithm base-10 of "wprob" will
be used as predictor in place of "wprob".</p></td>
    </tr>
    <tr>
      <th>classifier</th>
      <td><p>Classification model to use. Option "logistic" applies a
logistic regression model; option "lda" applies a Linear Discriminant
Analysis (LDA); "qda" applies a Quadratic Discriminant Analysis (QDA),
"pca.logistic" applies logistic regression model using the Principal
Component (PCs) estimated with Principal Component Analysis (PCA) as
predictor variables. pca.lda" applies LDA using PCs as predictor
variables, and the option "pca.qda" applies a Quadratic Discriminant
Analysis (QDA) using PCs as predictor variables.</p></td>
    </tr>
    <tr>
      <th>pval.col</th>
      <td><p>Column number for p-value used in the performance
analysis and estimation of the cutpoints. Default: NULL. If NULL it is
assumed that the column is named "wprob".</p></td>
    </tr>
    <tr>
      <th>n.pc</th>
      <td><p>Number of principal components (PCs) to use if the classifier is
not 'logistic'. In the current case, the maximun number of PCs is 4.</p></td>
    </tr>
    <tr>
      <th>center</th>
      <td><p>A logical value indicating whether the variables should be
shifted to be zero centered (same as in 'prcomp' prcomp). Only used if
classifier = "pcaLDA".</p></td>
    </tr>
    <tr>
      <th>scale</th>
      <td><p>A logical value indicating whether the variables should be
scaled to have unit variance before the analysis takes place (same as in
'prcomp' prcomp). Only used if classifier = "pcaLDA".</p></td>
    </tr>
    <tr>
      <th>interaction</th>
      <td><p>Variable interactions to consider in a logistic
regression model. Any pairwise combination of the variable "hdiv", "TV",
"wprob", and "pos" can be provided. For example: "hdiv:TV", "wprob:pos",
"wprob:TV", etc.</p></td>
    </tr>
    <tr>
      <th>output</th>
      <td><p>Type of output to request: output = c("conf.mat", "mc.val",
"boot.all", "all"). See below.</p></td>
    </tr>
    <tr>
      <th>prop</th>
      <td><p>Proportion to split the dataset used in the logistic regression:
group versus divergence (at DMPs) into two subsets, training and
testing.</p></td>
    </tr>
    <tr>
      <th>num.boot</th>
      <td><p>Number of bootstrap validations to perform in the evaluation
of the logistic regression: group versus divergence (at DMPs).</p></td>
    </tr>
    <tr>
      <th>num.cores, tasks</th>
      <td><p>Parameters for parallel computation using package
<code><a href='https://rdrr.io/pkg/BiocParallel/man/BiocParallel-package.html'>BiocParallel-package</a></code>: the number of cores to
use, i.e. at most how many child processes will be run simultaneously
(see <code><a href='https://rdrr.io/pkg/BiocParallel/man/bplapply.html'>bplapply</a></code> and the number of tasks per job
(only for Linux OS).</p></td>
    </tr>
    <tr>
      <th>seed</th>
      <td><p>Random seed used for random number generation.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>if TRUE, prints the function log to stdout</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>output = "conf.mat" will perform a logistic regression group versus
divergence (at DMPs) to evaluate the discrimination between
control-DMPs and treatment-DMPs. The evaluation of this classification
is provided through the function 'confusionMatrix' from R package
'caret'. "mc.val" will perform a 'num.boot'-times Monte Carlo
(bootstrap) validation and return a summary. By default function
'confusionMatrix' from R package caret' randomly splits the sample into
two subsets, training and testing, according to the supplied proportion
'prop' (i.e., prop = 0.6). After selecting output = "mc.val", the
function 'confusionMatrix' will be executed 'num.boot'-times, each time
performing a different random split of the sample. "boot.all" same as
"mc.val" plus a matrix with statistcs reported by 'confusionMatrix'.
"all" return a list with all the mentioned outputs.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The regulatory methylation signal is also an output from a natural
process that continuously takes place across the ontogenetic development
of the organisms. So, we expect to see methylation signal on natural
ordinary conditions. Here, to distinguish a control methylation signal
from a treatment, three classification models are provided: 1) logistic,
2) Linear Discriminant Analysis (LDA) and 3) Quadratic Discriminant
Analysis (QDA). In particular, four predictor variables can be used:
Hellinger divergence "hdiv", total variation "TV", probability of
potential DMP "wprob" and DMP genomic coordinated "pos". Principal
component analysis (PCA) is used to convert a set of observations of
possibly correlated predictor variables into a set of values of linearly
uncorrelated variables (principal components, PCs). The PCs are used as
new, uncorrelated predictor variables for LDA, QDA, and logistic
classifiers.</p><pre>A classification result with low accuracy and compromising values from
other classification performance indicators (see below) suggest that the
treatment does not induce a significant regulatory signal different
from control.
</pre>


    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span>(<span class='no'>cutpoint</span>, <span class='no'>PS</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"MethylIT"</span>)

<span class='co'>## DMPs are selected using the cupoints</span>
<span class='no'>DMPs</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='selectDIMP.html'>selectDIMP</a></span>(<span class='no'>PS</span>, <span class='kw'>div.col</span> <span class='kw'>=</span> <span class='fl'>9L</span>, <span class='kw'>cutpoint</span> <span class='kw'>=</span> <span class='no'>cutpoint</span>$<span class='no'>cutpoint</span>,
                   <span class='kw'>tv.cut</span> <span class='kw'>=</span> <span class='fl'>0.68</span>)

<span class='co'>## Classification of DMPs into two clases: DMPS from control and DMPs from</span>
<span class='co'>## treatment samples and evaluation of the classifier performance (for more</span>
<span class='co'>## details see ?evaluateDIMPclass).</span>
<span class='no'>perf</span> <span class='kw'>&lt;-</span> <span class='fu'>evaluateDIMPclass</span>(<span class='kw'>LR</span> <span class='kw'>=</span> <span class='no'>DMPs</span>,
                          <span class='kw'>column</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='kw'>hdiv</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>TV</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
                                     <span class='kw'>wprob</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>pos</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>),
                          <span class='kw'>classifier</span> <span class='kw'>=</span> <span class='st'>"lda"</span>, <span class='kw'>n.pc</span> <span class='kw'>=</span> <span class='fl'>4L</span>,
                          <span class='kw'>control.names</span> <span class='kw'>=</span>  <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"C1"</span>, <span class='st'>"C2"</span>, <span class='st'>"C3"</span>),
                          <span class='kw'>treatment.names</span> <span class='kw'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"T1"</span>, <span class='st'>"T2"</span>, <span class='st'>"T3"</span>),
                          <span class='kw'>center</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>scale</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
                          <span class='kw'>prop</span> <span class='kw'>=</span> <span class='fl'>0.6</span>)

<span class='co'>## Model classification performance</span>
<span class='no'>perf</span>$<span class='no'>Performance</span></div><div class='output co'>#&gt; Confusion Matrix and Statistics
#&gt; 
#&gt;           Reference
#&gt; Prediction  CT  TT
#&gt;         CT  62   0
#&gt;         TT   0 596
#&gt;                                      
#&gt;                Accuracy : 1          
#&gt;                  95% CI : (0.9944, 1)
#&gt;     No Information Rate : 0.9058     
#&gt;     P-Value [Acc &gt; NIR] : &lt; 2.2e-16  
#&gt;                                      
#&gt;                   Kappa : 1          
#&gt;                                      
#&gt;  Mcnemar's Test P-Value : NA         
#&gt;                                      
#&gt;             Sensitivity : 1.0000     
#&gt;             Specificity : 1.0000     
#&gt;          Pos Pred Value : 1.0000     
#&gt;          Neg Pred Value : 1.0000     
#&gt;              Prevalence : 0.9058     
#&gt;          Detection Rate : 0.9058     
#&gt;    Detection Prevalence : 0.9058     
#&gt;       Balanced Accuracy : 1.0000     
#&gt;                                      
#&gt;        'Positive' Class : TT         
#&gt;                                      </div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#details">Details</a></li>
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Robersy Sanchez.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


